Assets/MixedRealityToolkit.Services/InputSystem/NearInteractionTouchableUnityUI.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.Serialization;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Use a Unity UI RectTransform as touchable surface.
+    /// </summary>
+    [RequireComponent(typeof(RectTransform))]
+    public class NearInteractionTouchableUnityUI : BaseNearInteractionTouchable
```

Comment:
Doesn't this require a collider component too? If we require Unity UI to be wrapped in a collider for near interaction, we need to make sure that doesn't affect far interactions (i.e. interactions via raycast-based pointers).

Reply:
Background: Previously the BaseNearInteractionTouchable class was handling both the collider case and the collider-less case. The latter was only used for UnityUI case, so i've split the collider code off into a separate base class `ColliderNearInteractionTouchable` for simpler code.

So, no, the UnityUI case didn't and doesn't require a collider. There is a breaking change here: with the static list used previously the PokePointer would still have considered objects with UnityUI, but now they won't be included. However, even before this change the UnityUI case was quite pointless because it would never have generated events, since these would be raised through the FocusProvider with an additional raycast.

I can only assume that the UnityUI stuff is left over from HTK days and not functional. AFAIK we don't have any examples for it either.

Reply:
There are examples in HandInteractionExamples and HandInteractionUnityUIExample. They are functional and should be tested.

Reply:
I have no clue how the UnityUI thing works. The examples seem to be broken with this change, would be great to have an actual test indeed.

Reply:
Maybe check out `PokePointer.IsObjectPartOfTouchable`. It may shed some light on how the Unity UI stuff works.

Reply:
I don't know the details but I know that `MixedRealityInputModule` translates our pointer events into the mouse pointer events that Unity UI understands (see `GetMousePointerEventDataForMrtkPointer()`).

Reply:
So it looks like what is happening right now is that the PokePointer in `OnPreSceneQuery` [sets up a ray step](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/e18e7078bde38454816a0d3d2bd7c5e54bc2ccc5/Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs#L74), and the FocusProvider then [uses this to perform a graphics raycast](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/e18e7078bde38454816a0d3d2bd7c5e54bc2ccc5/Assets/MixedRealityToolkit.Services/InputSystem/FocusProvider.cs#L1078). This goes into the internal Unity system for handling UI.

The problem now is that there is no equivalent to the Physics.OverlapSphere function for the graphics/event system in Unity. We would need this to filter UI elements for the poke pointer, otherwise we'd have to go back to a static list for the UnityUI touchable variants ...

Reply:
I can't find a solution to this. If nobody has any idea we might as well drop this PR.

Reply:
I think we can do it, there's so many good things in this PR.

Reply:
I think that dispatching of input in UnityUI is implemented in [MixedRealityInputModule (MRIM)](https://github.com/julenka/MixedRealityToolkit-Unity/blob/e5dac4afce15cc6719b6cd11b965b1dcb0919a6a/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputModule.cs#L302).

Here's what I think happens (based on my experience with [this PR](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4765/files). 

1. MRIM tracks all active pointers by listening to [sourcedetected](https://github.com/julenka/MixedRealityToolkit-Unity/blob/e5dac4afce15cc6719b6cd11b965b1dcb0919a6a/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputModule.cs#L302), sourcelost
2. MRIM implements [IMRPointerHandler](https://github.com/julenka/MixedRealityToolkit-Unity/blob/e5dac4afce15cc6719b6cd11b965b1dcb0919a6a/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputModule.cs#L11)
3. OnPointerDown, MRIM updates state of its tracked pointers.
4. In [Process()](https://github.com/julenka/MixedRealityToolkit-Unity/blob/e5dac4afce15cc6719b6cd11b965b1dcb0919a6a/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputModule.cs#L115), MRIM actually translates the pointer states it is tracking (pressed, not pressed, etc) into instructions for Unity to ProcessMouseEvent.
5. As part of ProcessMouseEvent, Unity calls GetMousePointerEventData, which calls [GetMousePointerEventDataForMRTKPointer](https://github.com/julenka/MixedRealityToolkit-Unity/blob/e5dac4afce15cc6719b6cd11b965b1dcb0919a6a/Assets/MixedRealityToolkit.Services/InputSystem/MixedRealityInputModule.cs#L188). 
6. That method is the thing that has the meat -- it takes the position and rotation of the pointer (actually, the rotations of Rays[0]) and converts that into a ray. This ray then gets used to figure out which UI element is hit. We don't need to worry about this luckily -- Unity does its own casting to figure out which sub element was hit.

In short, I think as long as you can [set up that ray step](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/e18e7078bde38454816a0d3d2bd7c5e54bc2ccc5/Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Pointers/PokePointer.cs#L74) at some point so that the position of ray step matches some place in front of the surface, and the rotation points toward the surface (so that the ray will hit the Unity UI), you will be fine.

Reply:
Hm now reading Luis's comment above I feel like you and Luis probably already figured out what I said above...

Reply:
Here's an idea: In FindClosestTouchableForLayerMask, could you just do a graphics raycast for UnityUI? Don't worry about a sphere cast, just do a raycast. Reasoning is this: Unity UI is always going to be a flat rectangle, so you will always touch them from the front. You could set the position of the ray to be finger tip, and the direction of the ray the normal of UnityUI surface.

So, first find the colliders from volumes using the sphere cast, then also look for any UnityUI by doing a graphics raycast, ignoring all surfaces farther than touchableDistance.

Reply:
The problem is in the first link of this chain: finding all potential touchables to examine (i.e. the "broad phase"). Without this no RayStep will be created and no graphics raycast happens.

Doing a graphics raycast isn't a good solution, because i don't know at that point in which direction to cast the ray. Just casting to "forward" (what does that mean for a poke pointer?) will probably miss ui elements and cause problems down the line.

I'm considering just adding back the static list for just the Unity canvas objects, so we can at least get a reliable behavior. Then perhaps we can try to get Unity to supply a proper overlap test for their UI elements.

Reply:
In partner apps I've scene they usually just have one Unity UI canvas, maybe 2.

Reply:
Doesn't really matter if it's 2 or 1000 though, we'd need some global refs to the canvases to examine. Could be automatic registry in a static list as before, or a simple input system pointer to the canvas to use, but the latter is even more error-prone.

Reply:
I think static list for Unity UI is OK. It's definitely a case we want to still support (several partners depend on it).

Reply:
I've added back the static instances list for UnityUI touchables. It works fine in the examples, but would be great to have an automated test for this case too. I've tried creating a Canvas object and test components from script, but that didn't work for some reason (i'm not an expert on the UI system). Loading the prefab would be easier, but we haven't solved the question of test resources yet.

Reply:
Have a look at #5033 for an example of how to load prefabs. It isn't ideal because if the prefab names change the tests will break. But it will at least unblock you. Maybe a dictionary of common test resources would be good, or a single class that would load test prefabs. This way there's just one place where the paths to prefabs used for tests exist. Something like :